### PR TITLE
Nightmare Inserts in Unit Testing & Upgrades to Subsystems

### LINKED ISSUE #1872: Napthal Fuel bad del on blackcat?
### Testmerges

1760, 1841, 955, 1707

### Description of the bug

```
runtime error: bad del
proc name: qdel (/proc/qdel)
  source file: garbage.dm,264
  usr: (src)
  src: null
  call stack:
qdel(null, 0)
UT-Napthal Fuel (/datum/reagent/napalm/ut): Destroy(0)
qdel(UT-Napthal Fuel (/datum/reagent/napalm/ut), 0)
Garry (/mob/living/simple_animal/cat/Runtime): Destroy(0)
Garry (/mob/living/simple_animal/cat/Runtime): Destroy(0)
qdel(Garry (/mob/living/simple_animal/cat/Runtime), 0)
the black cat (/obj/item/holder/blackcat): Destroy(0)
the black cat (/obj/item/holder/blackcat): Destroy(0)
the black cat (/obj/item/holder/blackcat): Destroy(0)
the black cat (/obj/item/holder/blackcat): Destroy(0)
...
Unknown (/mob/living/carbon/human): Destroy(0)
Unknown (/mob/living/carbon/human): Destroy(0)
Unknown (/mob/living/carbon/human): Destroy(0)
qdel(Unknown (/mob/living/carbon/human), 0)
the spawn pool (/obj/effect/alien/resin/special/pool): melt body(0)
/datum/callback (/datum/callback): InvokeAsync()
Timer (/datum/controller/subsystem/timer): fire(0)
Master (/datum/controller/master): RunQueue()
Master (/datum/controller/master): StartProcessing(0)
```

### What's the difference with what should have happened?

Not runtime

### How do we reproduce this bug?

???

### Issue Bingo

- [ ] Issue could be reproduced at least once
- [X] Issue happened in a recent (less than 7 days ago) round
- [X] Couldn't find an existing issue about this (https://github.com/cmss13-devs/cmss13/issues)

### LINKED ISSUE #3017: Multiple errors in flaky test create_and_destroy
Flaky tests were detected in [this test run](https://github.com/cmss13-devs/cmss13/actions/runs/4614127118/attempts/1). This means that there was a failure that was cleared when the tests were simply restarted.
Failures:
```
create_and_destroy: /obj/item/device/flashlight/flare/signal/debug hard deleted 1 times out of a total del count of 1 at code/modules/unit_tests/create_and_destroy.dm:104
create_and_destroy: /datum/cas_signal hard deleted 1 times out of a total del count of 1 at code/modules/unit_tests/create_and_destroy.dm:104
create_and_destroy: /obj/item/device/rcd hard deleted 2 times out of a total del count of 2 at code/modules/unit_tests/create_and_destroy.dm:104
create_and_destroy: /obj/item/device/rcd/borg hard deleted 2 times out of a total del count of 2 at code/modules/unit_tests/create_and_destroy.dm:104
create_and_destroy: /mob/living/silicon hard deleted 1 times out of a total del count of 3 at code/modules/unit_tests/create_and_destroy.dm:104
create_and_destroy: /obj/item/reagent_container/food/drinks/cans/beer hard deleted 1 times out of a total del count of 15 at code/modules/unit_tests/create_and_destroy.dm:104
create_and_destroy: /obj/item/explosive/grenade/flashbang/cluster/segment hard deleted 1 times out of a total del count of 1 at code/modules/unit_tests/create_and_destroy.dm:104
create_and_destroy: /obj/item/explosive/grenade/flashbang/cluster_piece hard deleted 1 times out of a total del count of 1 at code/modules/unit_tests/create_and_destroy.dm:104
create_and_destroy: /obj/item/tank/jetpack hard deleted 1 times out of a total del count of 1 at code/modules/unit_tests/create_and_destroy.dm:104
create_and_destroy: /obj/item/tank/jetpack/void hard deleted 2 times out of a total del count of 2 at code/modules/unit_tests/create_and_destroy.dm:104
create_and_destroy: /obj/item/tank/jetpack/oxygen hard deleted 1 times out of a total del count of 1 at code/modules/unit_tests/create_and_destroy.dm:104
create_and_destroy: /obj/item/tank/jetpack/carbondioxide hard deleted 1 times out of a total del count of 1 at code/modules/unit_tests/create_and_destroy.dm:104
create_and_destroy: /obj/structure/machinery/computer/supplycomp/vehicle hard deleted 1 times out of a total del count of 1 at code/modules/unit_tests/create_and_destroy.dm:104
create_and_destroy: /obj/structure/machinery/alarm hard deleted 1 times out of a total del count of 1 at code/modules/unit_tests/create_and_destroy.dm:104
create_and_destroy: /obj/effect/landmark/costume/madscientist hard deleted 1 times out of a total del count of 1 at code/modules/unit_tests/create_and_destroy.dm:104
create_and_destroy: /obj/effect/landmark/costume/scratch hard deleted 1 times out of a total del count of 1 at code/modules/unit_tests/create_and_destroy.dm:104
create_and_destroy: /obj/effect/landmark/costume/highlander hard deleted 1 times out of a total del count of 1 at code/modules/unit_tests/create_and_destroy.dm:104
```

### PR DESCRIPTION
**Pull request by Drulikar**
# About the pull request

This PR is primarily an attempt to fix #3017 (or at least make it more consistent) and to just get nightmare inserts actually tested.

During testing, nightmare inserts now skip all probability and conditional checks to force all to occur (overlapping inserts might be problematic in the future - remove commented blocks if so - but doesn't currently seem to be a problem).

The nightmare subsystem was also not running at all when UNIT_TESTS was defined because HandleTestRun called request_start which calls nightmare.prepare_game prior to nightmare.Initialize. That was making the nightmare subsystem think there were no nightmare inserts to even attempt to spawn. Now for HandleTestRun, it will wait UNTIL SSticker is initialized which should include nightmare initialization and help with making create_and_destroy testing more reliable.

The map linter also now cares about a table placed on a table (because the latter will just delete the former in Initialize was causing a bad qdel hint). Also, datum/Destroy will now create no sleeping lints.

This PR also adds the reagents subsystem to handle setting up the global lists for chemistry. This was done to fix the napalm bad delete runtime on shiva where their fire reagents were never properly initialized. Previously the first reagents datum Initialize was needed to trigger initialization)

Since the tests still ended up flaky, this PR now partially ports several TG PRs:
- https://github.com/tgstation/tgstation/pull/64705
- https://github.com/tgstation/tgstation/pull/69742
- https://github.com/tgstation/tgstation/pull/72033
- https://github.com/tgstation/tgstation/pull/72401
- https://github.com/tgstation/tgstation/pull/73903
- https://github.com/tgstation/tgstation/pull/74604
- https://github.com/tgstation/tgstation/pull/75535
- https://github.com/tgstation/tgstation/pull/71177
- https://github.com/tgstation/tgstation/pull/75600

A side effect of the last PR in above seems to unfortunately make it so usually only one hard delete will be detected at a time. There will also likely be future improvements that TG makes to the atoms subsystem because of weird edge cases ever multiple maps load simultaneously.

The only edge case I have experienced for us is late_loaders filling up while it is being dequeued, but the loop has been altered to accommodate this to check the length every iteration. Some launches of the game may spit out debug messages like `We started maploading while we were already maploading`. TG needs to sort out how they want to solve this edge case, but it does not appear to cause any problems for us.

For live using the flag `-DEXPERIMENT_515_DONT_CACHE_REF` will likely be a benefit, but try it both with and without it. The hardref experiment I have disabled even for UNIT_TESTING at the moment.

# Explain why it's good for the game

- Less hard deletes and runtimes including fix #1872
- More testing and lints
- Less flaky hard delete testing hopefully

# Changelog
:cl: Drathek
code: Added unit_testing to annotate and fail for runtimes, added a unit_test for 515, and improved debug information for reference lookups.
code: Moved chemistry list setup to a new reagents subsystem rather than potentially before every reagents initialization.
code: Ported many updates to SSgarbage, SSatoms, and create_and_destroy fixing inconsistent garbage handling, an edge case with atoms late loading, and improved speed.
fix: Fix nightmare inserts not occurring during unit testing (they will also ignore probabilities for testing now).
fix: Fix runtimes for bad deletes and various hard deletes.
maptweak: Fix some nightmare inserts causing errors.
server: Initialize log for errors/warnings for SSatoms is now in each round's respective log folder.
/:cl:

### CODE REVIEW

**Comment by fira** [comment]
Well that's a very good idea

**Comment by Drulikar** [comment]
Let me know if you think it fixes the flaky test; I'm not quite convinced yet since create_and_destroy has still been finicky about what it even thinks is wrong even through out the development of this PR - and often times only on a single map. Anything that it personally instantiates in its own test, it should always report those errors - my hope was making nightmares consistent would make it consistent...

File: code/modules/shuttle/shuttles/dropship.dm [hunk 1]
```diff
@@ -148,6 +148,13 @@
 	. = ..()
 	link_landing_lights()
 
+/obj/docking_port/stationary/marine_dropship/Destroy()
+	. = ..()
+	for(var/obj/structure/machinery/landinglight/light in landing_lights)
+		light.linked_port = null
+	landing_lights.Cut()
+	landing_lights = null // We didn't make them, so lets leave them
```
Comment (fira) [inline code/modules/shuttle/shuttles/dropship.dm:156]: FWIW Cut + Nulling are redundant

**Review by fira — APPROVED** [review]

File: code/modules/shuttle/shuttles/dropship.dm (hunk 1 above)
Comment (Drulikar) [reply code/modules/shuttle/shuttles/dropship.dm:156]: If so then why do the qdel defines do so? https://github.com/cmss13-devs/cmss13/blob/60248f5e5d31b7091e98fabf117cee21634fcc9a/code/__DEFINES/qdel.dm#L52

File: code/controllers/subsystem/reagents.dm [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+			else if(isNegativeProperty(prop))
+				chemical_properties_list["negative"][prop.name] = prop
+			else if(isNeutralProperty(prop))
+				chemical_properties_list["neutral"][prop.name] = prop
+			else if(isPositiveProperty(prop))
+				chemical_properties_list["positive"][prop.name] = prop
+
+/datum/controller/subsystem/reagents/proc/prepare_reagents()
+	//I dislike having these here but map-objects are initialised before world/New() is called. >_>
+	set waitfor = 0
```
Comment (harryob) [inline code/controllers/subsystem/reagents.dm:50]:
```suggestion
	set waitfor = FALSE
```

**Review by harryob — CHANGES_REQUESTED** [review]

File: code/controllers/subsystem/reagents.dm (hunk 2 above)
Comment (Drulikar) [reply code/controllers/subsystem/reagents.dm:50]: TBH I don't know if that set is still even necessary or needed, but the original proc had it and I didn't notice any issues.

**Comment by Drulikar** [comment]
Based on https://github.com/cmss13-devs/cmss13/actions/runs/5003373497/jobs/8964728495?pr=3305 this is still unfortunately flaky, so there's still something wrong with SSgarbage not recording issues correctly, or create_and_destroy not reliably prompting issues.

**Comment by Drulikar** [comment]
Can confirm the UNTIL change fixed the issue with nightmares not loading for a unit_test, but create_and_destroy is where it should be.
(Left: two runs of the commit before this PR, Right: two runs of the latest commit)
image

**Review by harryob — APPROVED** [review]

**Pull request merged**